include/zephyr/drivers/comparator.h

```diff
@@ -0,0 +1,201 @@
+/*
+ * Copyright (c) 2023 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#ifndef ZEPHYR_INCLUDE_DRIVERS_COMPARATOR_H_
+#define ZEPHYR_INCLUDE_DRIVERS_COMPARATOR_H_
+
+/**
+ * @brief Comparator Driver APIs
+ * @defgroup comparator_interface Comparator Driver APIs
+ * @ingroup io_interfaces
+ * @{
+ */
+
+#include <zephyr/device.h>
+#include <zephyr/dt-bindings/comparator/comparator.h>
+#include <zephyr/kernel.h>
+#include <zephyr/sys/util_macro.h>
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+/**
+ * @name Comparator state flags
+ * @{
+ */
+
+/** Positive input voltage is below the negative input voltage. */
+#define COMPARATOR_STATE_BELOW BIT(0)
+/** Positive input voltage is above the negative input voltage. */
+#define COMPARATOR_STATE_ABOVE BIT(1)
+
+/** @cond INTERNAL_HIDDEN */
+#define COMPARATOR_STATE_RESULT_MASK  0x3UL
+/* Position in the state bitfield from which vendor-specific flags can be defined. */
+#define COMPARATOR_STATE_VENDOR_POS 8
+/** @endcond */
+
+/** @} */
+
+/**
+ * @brief Comparator callback.
+ *
+ * @param dev Comparator device instance.
+ * @param state Bitfield that specifies the reported state. This parameter can
+ *              contain COMPARATOR_STATE_* values and vendor-specific flags.
+ * @param user_data User data.
+ */
+typedef void (*comparator_callback_t)(const struct device *dev,
+				      uint32_t state,
```

Comment:
nit: all bit fields should be typedeffed (e.g. `comparator_state_t`), so we can make future width changes without breaking.